src/summary_statistics/mod.rs

```diff
@@ -28,13 +28,17 @@ where
     where
         A: Clone + FromPrimitive + Add<Output = A> + Div<Output = A> + Zero;
 
-    /// Returns the [`arithmetic weighted mean`] x̅ of all elements in the array. Assumes that the
-    /// weights are already normalized.
+    /// Returns the [`arithmetic weighted mean`] x̅ of all elements in the array. Use `weighted_sum`
```

Comment:
What happens if the weights sum to `0`? I assume the function panics:
- we should add it to the docs;
- we should add a test for it.

Reply:
Oh, I forgot about this. Thank you.

Reply:
In fact, it panics on integer and it returns a NaN on floating point, which is quite normal. Do we want to panic on floating point numbers or we let the users learn the hard facts of IEEE754?

Reply:
I would let them know the hard facts, unfortunately :joy: 
We could use something along the same lines of what we say for `mean_axis` in `ndarray`: 

> Panics if axis is out of bounds or if the length of the axis is zero and division by zero panics for type A.

Which makes me think that we need to add a test and a note about zero-length axis on both `weighted_mean_axis` and `weighted_sum_axis`.

Reply:
Sorry, I don't understand this part. How can "the length of the axis" be zero? Are we talking about a `Array0` (single number)? Surely the len of an `Array0` is 1 and not 0? Or it's considered a data **point** and it has no length?

Also, reading `ndarray`'s version, it seems like it's saying "Panics if x OR (y AND z)" Should it be all "or"?

Reply:
Sorry to disturb you again, but I'm lost here.

1. Those zero-length dimensions don't make sense to me. It looks like they are uselessly complex empty arrays. Do you have more information on them? Because I can't code something for a "feature" that I don't understand.
2. I removed the `* MultiInputError::EmptyInput if self is empty` error on `weighted_mean_axis` and `weighted_sum_axis` because I don't understand why they should return an error when the array is empty. It does seem to make sense for `weighted_mean_axis` because we divide all lanes by weights_sum, but there's no lane to divide anyway.
3. I can't respect `Panics if axis is out of bounds or if the length of the axis is zero and division by zero panics for type A.` because my functions use `map_axis`, which panics with an error not defined in [`map_axis` documentation](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.map_axis).
```
let a = Array2::<f32>::zeros((0, 20));
a.mean_axis(Axis(0)); // pass
a.weighted_mean_axis(Axis(0), &Array1::zeros(0)).unwrap(); // fail
    panicked at 'collapse_axis: Index 0 must be less than axis length 0 for array with shape [0, 20]
```

Of course I could try to use something else (there are other ways to code `weighted_sum_axis`), but I don't see why I should, mostly because of 1).

Reply:
No issue at all, happy to reach consensus on these changes :smile: 

> 1. Those zero-length dimensions don't make sense to me. It looks like they are uselessly complex empty arrays. Do you have more information on them? Because I can't code something for a "feature" that I don't understand.

I wouldn't indeed consider them a feature - they are edge cases (empty arrays with associated non-trivial dimension-related information), but given that we are writing a library we need to take care of them (or at least document what happens if you pass them in).
Would it possible to forbid them? It's a possibility, but as long as `ndarray`'s codebase allows them, we need to deal with them. 

> 2. I removed the `* MultiInputError::EmptyInput if self is empty` error on `weighted_mean_axis` and `weighted_sum_axis` because I don't understand why they should return an error when the array is empty. It does seem to make sense for `weighted_mean_axis` because we divide all lanes by weights_sum, but there's no lane to divide anyway.

I agree on `weighted_sum_axis` / `weighted_sum` - there is a very good default value to return (`A::zero()`) and the behaviour is consistent with [`sum`](https://docs.rs/ndarray/0.13.0/ndarray/struct.ArrayBase.html#method.sum) in `ndarray`.
For the mean though, it gets trickier. What is the mean of an empty array of integers? `0`? If would argue that there is no sound answer to that question, hence it makes sense to return `None`. Given that those functions can fail in multiple ways (mismatched shapes), it's more ergonomic to return it as an error variant than a `Result<Option<A>, ErrType>` (see [`mean`](https://docs.rs/ndarray/0.13.0/ndarray/struct.ArrayBase.html#method.mean) in `ndarray`).

> 3. I can't respect `Panics if axis is out of bounds or if the length of the axis is zero and division by zero panics for type A.` because my functions use `map_axis`, which panics with an error not defined in [`map_axis` documentation](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.map_axis).
> 
> ```
> let a = Array2::<f32>::zeros((0, 20));
> a.mean_axis(Axis(0)); // pass
> a.weighted_mean_axis(Axis(0), &Array1::zeros(0)).unwrap(); // fail
>     panicked at 'collapse_axis: Index 0 must be less than axis length 0 for array with shape [0, 20]
> ```
> 
> Of course I could try to use something else (there are other ways to code `weighted_sum_axis`), but I don't see why I should, mostly because of 1).

This was a bug in `ndarray:0.12.1` (indeed when dealing with `0`-length axis :stuck_out_tongue:). It has been fixed in `0.13` - as soon as the CI is finished in https://github.com/rust-ndarray/ndarray-stats/pull/52 I will merge it into `master`, so that you can get the fix in this branch as well.

Reply:
https://github.com/rust-ndarray/ndarray-stats/pull/52 has been merged - I'd say that we could this PR merged and then cut a release :+1: